driver.go

```diff
@@ -184,10 +183,10 @@ func (d *DockerOnTop) Mount(request *volume.MountRequest) (*volume.MountResponse
 	}
 	defer activemountsdir.Close() // There is nothing I could do about the error (logging is performed inside `Close()` anyway)
 
-	_, readDirErr := activemountsdir.ReadDir(1) // Check if there are any files inside activemounts dir
-	if errors.Is(readDirErr, io.EOF) {
-		// No files => no other containers are using the volume. Need to mount the overlay
-
+	mount, err := d.Activate(request, activemountsdir)
+	if err != nil {
+		return nil, internalError("failed to activate the active mount:", err)
+	} else if mount {
 		lowerdir := thisVol.BaseDirPath
 		upperdir := d.upperdir(request.Name)
 		workdir := d.workdir(request.Name)
```

Comment:
Read [this comment](https://github.com/kolayne/docker-on-top/pull/20/files#r1693985907) first.

<br>

Here you, unfortunately, did kind of the opposite (similarly, you reversed the original order, but it has the inverse effect in this case).

Combining the logic inside `d.Activate` and the code _below_ this comment up to and including the `syscall.Mount` (that's the code you did not modify), the overall (new) logic is as follows:

- Check if the volume is already mounted based on `activemount` files. If it's not:
- Create a new `activemount` file to indicate that I am the user of the volume \[may fail\]
- Perform the actual mount of the `overlay` \[may fail\]

Now, imagine the file creation step succeeds but the next one fails. For that container, of course, we'll report the error and it will refuse to start, but the volume ends up in a _very bad_ state. If another container tries to use that same volume, it will perform the same steps above, but in the first one it will find an activemount file from the previous container and will assume that the volume is already mounted, while it in fact is not!!

The reverse order is not very good either but slightly better: if the first container manages to mount the volume but doesn't manage to mark it as mounted, it, likewise, won't let the current container start (so that `overlay` mount remains unused), then, when the next container attempts a mount, it will run the same steps and, in the trickiest case, will just mount the same `overlay` on top of the old one and start using it (which is weird because there are two `overlay`s mounted simultaneously using the same internal directories, but should be ok, as the first overlay is never used, being shadowed by the new one).

<hr>

So,

1. First possible improvement for your code: your implementation should at least emit a scary screaming error! (see [this example](https://github.com/kolayne/docker-on-top/blob/1bfffce2ff8960b549a9d5fb0a894895163e5ca6/driver.go#L301-L309) from the old code - this error was used in a very similar situation)
2. Second, it would be much better if you could fix things such that the order of the system calls (mount/unmount call and activemount file creation/deletion call) is old (as in my code) for the `Mount` operation but is new (as in your code) for the `Unmount` operation.

In fact, I now see how it's possible to completely fix the problem in the `Mount` part, so that we never even attempt to mount `overlay` on top of anything old. Of course, this is out of scope of this PR, so I'll create a separate issue (but it still requires that you restore the old order here in `Mount`)

Reply:
I have improved the way Mount function works, I keep the new order but in case the filesystem mount fails it will undo the `activemount` creation so it returns the function in a consistent state: filesystem not mounted and no activemount present.

01de48d2fe2c0cfc4fb01f1fd34247ca88a11e98